lib/spack/spack/test/cmd/mirror.py

```diff
@@ -288,3 +295,115 @@ def test_mirror_destroy(
 
     uninstall("-y", spec_name)
     mirror("remove", "atest")
+
+
+@pytest.mark.usefixtures("mock_packages")
+class TestMirrorCreate(object):
+    @pytest.mark.regression("31736", "31985")
+    def test_all_specs_with_all_versions_dont_concretize(self):
+        args = MockMirrorArgs(exclude_file=None, exclude_specs=None)
+        specs = spack.cmd.mirror.all_specs_with_all_versions(args)
+        assert all(not s.concrete for s in specs)
+
+    @pytest.mark.parametrize(
+        "cli_args,error_str",
+        [
+            # Passed more than one among -f --all and specs
+            ({"specs": "hdf5", "file": None, "all": True}, "cannot specify specs on command line"),
+            (
+                {"specs": None, "file": "input.txt", "all": True},
+                "cannot specify specs with a file if",
+            ),
+            (
+                {"specs": "hdf5", "file": "input.txt", "all": False},
+                "cannot specify specs with a file AND",
+            ),
+            ({"specs": None, "file": None, "all": False}, "no packages were specified"),
+            # Passed -n along with --all
+            (
+                {"specs": None, "file": None, "all": True, "versions_per_spec": 2},
+                "cannot specify '--versions_per-spec'",
```

Comment:
Because I took the help message as the declared intent of what the function should do:
```console
  -a, --all             mirror all versions of all packages in Spack, or all packages in the current environment if there is an active environment (this requires significant time and space)
```
If we want to have something that collects only the latest version, can we add an argument like `--all-latest` to do that?

Reply:
Also, can we do this in a follow up PR?

Reply:
Yes. If I'm the only one who's ever used it, I don't mind if it takes a while to come back, it doesn't look hard to reintroduce (i.e. it doesn't look like any of the changes in this PR make it difficult to extract a limited number of versions for each package).